File: csharp/TheFactory.Datastore/src/MemoryTablet.cs
using System;
using System.Threading;
using System.Linq;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace TheFactory.Datastore {

    public class MemoryTablet : ITablet {
        private OurSortedDictionary<Slice, Slice> backing;
        private AsyncLock backingLock = new AsyncLock();

        public int ApproxSize { get; private set; }

        public string Filename {
            get {
                return null;
            }
        }

        // Deleted key marker -- Get() and the Find() enumerator
        // should check against this reference and set KV.IsDeleted
        // when appropriate.
        public static Slice Tombstone = (Slice)(new byte[] {0x74, 0x6f, 0x6d, 0x62});

        public MemoryTablet() {
            backing = new OurSortedDictionary<Slice, Slice>(new KeyComparer());
            ApproxSize = 0;
        }

        public async Task Apply(Batch batch) {
            if (batch.IsEmpty()) {
                return;
            }

            foreach (IKeyValuePair kv in batch.Pairs()) {
                if (kv.IsDeleted) {
                    await Delete(kv.Key.Detach());
                    ApproxSize += kv.Key.Length;
                } else {
                    await Set(kv.Key.Detach(), kv.Value.Detach());
                    ApproxSize += kv.Key.Length + kv.Value.Length;
                }
            }
        }

        public async Task Set(Slice key, Slice val) {
            using (await backingLock.LockAsync()) {
                backing[key] = val;
            }
        }

        public async Task Delete(Slice key) {
            await Set(key, Tombstone);
        }

        public void Close() {
            backing.Clear();
        }

        public IAsyncEnumerable<IKeyValuePair> Find(Slice term) {
            if (backing.Count == 0) {
                return AsyncEnumerable.Empty<IKeyValuePair>();
            }

            return AsyncEnumerableEx.Return(() => backingLock.LockAsync())
                .SelectMany(l => {
                    IEnumerator<KeyValuePair<Slice, Slice>> items;

                    if (term == null) {
                        items = backing.GetEnumerator();
                    } else {
                        // skip all elements less than term
                        items = backing.GetSuffixEnumerator(term);
                    }

                    return AsyncEnumerable.Generate(new Pair(), _ => {
                        if (items.MoveNext()) return true;

                        l.Dispose();
                        return false;
                    }, ret => {
                        var kv = items.Current;

                        ret.Reset();
                        ret.Key = kv.Key;
                        if (ReferenceEquals(kv.Value, Tombstone)) {
                            ret.IsDeleted = true;
                        } else {
                            ret.Value = kv.Value;
                        }

                        return ret;
                    }, x => x);
                });
        }

        private class KeyComparer : IComparer<Slice> {
            public int Compare(Slice x, Slice y) {
                return Slice.Compare(x, y);
            }
        }
    }

}

File: java/src/main/java/com/thefactory/datastore/MemoryTablet.java
package com.thefactory.datastore;

import java.util.Comparator;
import java.util.concurrent.ConcurrentSkipListMap;
import java.util.Iterator;
import java.util.Map;
import java.util.NoSuchElementException;

public class MemoryTablet {    
    private final ConcurrentSkipListMap<Slice, Slice> backing;
    private long size = 0;

    public static Slice tombstone = new Slice(new byte[] {(byte)0x74, (byte)0x6f, (byte)0x6d, (byte)0x62});

    public MemoryTablet() {
        backing = new ConcurrentSkipListMap<Slice, Slice>(
            new Comparator<Slice>() {
                public int compare(Slice x, Slice y) {
                    return Slice.compare(x, y);
                }
            }
        );        
    }

    public void set(Slice key, Slice value) {
        backing.put(key, value);
        synchronized(this) {
            size += key.getLength() + value.getLength();
        }
    }

    public void delete(Slice key) {
        set(key, tombstone);
    }

    public void close() {
        backing.clear();
    }

    public Iterator<KV> find() {
        return find(null);
    }

    public Iterator<KV> find(final Slice term) {
        if (backing.size() == 0) {
            return empty();
        }
        return new Iterator<KV>() {
            Iterator<Map.Entry<Slice, Slice>> itemIterator;
            {
                if (term == null){
                    itemIterator = backing.entrySet().iterator();
                } else {
                    itemIterator = backing.tailMap(term, true).entrySet().iterator();
                }

            }

            public boolean hasNext() {
                return itemIterator.hasNext();
            }

            public KV next() {
                Map.Entry<Slice, Slice> item = itemIterator.next();
                KV ret = new KV();
                if(item.getValue() == tombstone){
                    ret.tombstone(item.getKey());
                } else {
                    ret.reset(item.getKey(), item.getValue());
                }
                return ret;
            }

            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }

    public void apply(Batch batch) {
        if (batch.isEmpty()) {
            return;
        }

        Iterator<KV> kvs = batch.pairs();
        while(kvs.hasNext()) {
            KV kv = kvs.next();
            if(kv.isDeleted()) {
                delete(kv.getKey().detach());
            } else {
                set(kv.getKey().detach(), kv.getValue().detach());
            }
        }
    }

    public long size() {
        return size;
    }

    private Iterator<KV> empty() {
        return new Iterator<KV>() {
            public boolean hasNext() {
                return false;
            }
            
            public KV next() {
                throw new NoSuchElementException();
            }
                
            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }
}
